File: src/main/java/xf/xflp/opt/construction/multitype/OneContainerNTypeAddPacker.java
package xf.xflp.opt.construction.multitype;

import xf.xflp.base.XFLPModel;
import xf.xflp.base.container.Container;
import xf.xflp.base.item.Item;
import xf.xflp.exception.XFLPException;
import xf.xflp.opt.Packer;
import xf.xflp.opt.construction.strategy.Strategy;

import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

/**
 * Copyright (c) 2012-2022 Holger Schneider
 * All rights reserved.
 *
 * This source code is licensed under the MIT License (MIT) found in the
 * LICENSE file in the root directory of this source tree.
 *
 *
 * This packer puts the items in a sequence for each container type into single container.
 * Items will only be added to a container.
 *
 * @author hschneid
 *
 */
public class OneContainerNTypeAddPacker implements Packer {

	@Override
	public void execute(XFLPModel model) throws XFLPException {
		Strategy strategy = model.getParameter().getPreferredPackingStrategy();

		// Create one container per type
		List<Container> containers = getContainers(model);
		// Try to insert items in containers
		List<Item> unplannedItems = new MultiBinAddHeuristic(strategy, model.getStatusManager())
				.createLoadingPlan(Arrays.asList(model.getItems()), containers);

		// Put result into model
		model.setContainers(containers.toArray(new Container[0]));
		model.setUnplannedItems(unplannedItems.toArray(new Item[0]));
	}

	private List<Container> getContainers(XFLPModel model) {
		return Arrays.stream(model.getContainerTypes())
				.map(Container::newInstance)
				.collect(Collectors.toList());
	}
}

File: src/xflp/opt/construction/multitype/one_container_n_type_add_packer.py
# Copyright (c) 2012-2025 Holger Schneider
# All rights reserved.
#
# This source code is licensed under the MIT License (MIT) found in the
# LICENSE file in the root directory of this source tree.

"""
This packer puts the items in a sequence for each container type into single container.
Items will only be added to a container.

Author: hschneid
"""

from typing import List, TYPE_CHECKING

from xflp.opt.packer import Packer
from xflp.opt.construction.multitype.multi_bin_add_heuristic import MultiBinAddHeuristic

if TYPE_CHECKING:
    from xflp.base.xflp_model import XFLPModel
    from xflp.base.container.container import Container


class OneContainerNTypeAddPacker(Packer):
    """Packer for one container per type (add only)."""

    def execute(self, model: 'XFLPModel') -> None:
        """
        Execute the packer.

        Args:
            model: The XFLP model

        Raises:
            XFLPException: If packing fails
        """
        strategy = model.get_parameter().get_preferred_packing_strategy()

        # Create one container per type
        containers = self._get_containers(model)

        # Try to insert items in containers
        unplanned_items = MultiBinAddHeuristic(
            strategy,
            model.get_status_manager(),
            model.get_parameter()
        ).create_loading_plan(
            list(model.get_items()),
            containers
        )

        # Put result into model
        model.set_containers(containers)
        model.set_unplanned_items(unplanned_items)

    def _get_containers(self, model: 'XFLPModel') -> List['Container']:
        """Create one container per type."""
        return [ct.new_instance() for ct in model.get_container_types()]
